toolchain/semantics/semantics_builtin_kind.def

```diff
@@ -6,30 +6,56 @@
 //
 // It does not use `#include` guards, and instead is designed to be `#include`ed
 // after the x-macro is defined in order for its inclusion to expand to the
-// desired output. The x-macro for this header is `CARBON_PARSE_NODE_KIND`. The
-// definition provided will be removed at the end of this file to clean up.
+// desired output.
+//
+// x-macros come in two forms:
+//   CARBON_SEMANTICS_BUILTIN_KIND_NAME(Name)
+//     Used as a fallback if other macros are missing.
+//   CARBON_SEMANTICS_BUILTIN_KIND(Name, Type)
+//     Defines a builtin kind with the associated type, which must also be
+//     builtin.
+//
+// Note that Invalid is provided with CARBON_SEMANTICS_BUILTIN_KIND_NAME but not
+// CARBON_SEMANTICS_BUILTIN_KIND.
+
+#if !(defined(CARBON_SEMANTICS_BUILTIN_KIND_NAME) || \
+      defined(CARBON_SEMANTICS_BUILTIN_KIND))
+#error \
+    "Must define CARBON_SEMANTICS_BUILTIN_KIND family x-macros to use this file."
+#endif
+
+// If CARBON_SEMANTICS_BUILTIN_KIND_NAME is undefined, ignore calls.
+#ifndef CARBON_SEMANTICS_BUILTIN_KIND_NAME
+#define CARBON_SEMANTICS_BUILTIN_KIND_NAME(Name)
+#endif
 
+// If CARBON_SEMANTICS_BUILTIN_KIND is undefined, delegate calls to
+// CARBON_SEMANTICS_BUILTIN_KIND_NAME.
 #ifndef CARBON_SEMANTICS_BUILTIN_KIND
-#error "Must define the x-macro to use this file."
+#define CARBON_SEMANTICS_BUILTIN_KIND(Name, ...) \
+  CARBON_SEMANTICS_BUILTIN_KIND_NAME(Name)
 #endif
 
 // Tracks expressions which are valid as types.
-CARBON_SEMANTICS_BUILTIN_KIND(TypeType)
+// This has a deliberately self-referential type.
+CARBON_SEMANTICS_BUILTIN_KIND(TypeType, TypeType)
 
 // Used when a SemanticNode has an invalid type, which should then be ignored
 // for future type checking.
-CARBON_SEMANTICS_BUILTIN_KIND(InvalidType)
+// This has a deliberately self-referential type.
+CARBON_SEMANTICS_BUILTIN_KIND(InvalidType, InvalidType)
```

Comment:
Why does this use `InvalidType` not `TypeType` as its type? If we have an invalid type, I would have expected that we know that it is a type, so in principle we should get better error recovery if we model the type as `type`. Is it better to use `InvalidType` here?

(I see this is unchanged from the previous implementation, so don't treat this question as blocking for this PR.)

Reply:
Because it's invalid. :)

Put differently, if I write the code:

```
var x: auto = invalid;
```

Here, the type of the the value `invalid` is `InvalidType`.

Does that mean that the `auto` should be `InvalidType`, or `TypeType`?

That is, if I now write code:

```
var x: auto = invalid;
x.Foo();
```

If I say that `auto` is `TypeType`, do I get an error that `Foo()` doesn't exist on `type`? If I use `auto` is `InvalidType`, then I can suppress an error.

Reply:
I would expect that the type of `x` would be deduced as the type of the right-hand side, which is `InvalidType`, but the *type of* the type of `x` would always be `type`.

Reply:
Let's discuss this in person, I think this would probably lead to unnecessary diagnostic errors in generics but maybe I'm missing something.